homeassistant/components/bmw_connected_drive/quality_scale.yaml

```diff
@@ -0,0 +1,80 @@
+rules:
```

Comment:
I do not think separate entities make sense here - they are just describing the more important state of "is there a problem" and will not change very often (maybe twice a year).
I feel it rather clutters the integration than being useful.

Reply:
There are quite a lot of them like
```py
        attr_fn=lambda v, u: {
            "door_lock_state": v.doors_and_windows.door_lock_state.value
        },
```
door lock state has the following options
```py
class LockState(StrEnum):
    """Possible states of the door locks."""

    LOCKED = "LOCKED"
    SECURED = "SECURED"
    SELECTIVE_LOCKED = "SELECTIVE_LOCKED"
    PARTIALLY_LOCKED = "PARTIALLY_LOCKED"
    UNLOCKED = "UNLOCKED"
    UNKNOWN = "UNKNOWN"
```
In theory this would be a very very nice ENUM device class sensor. We could disable it by default, but it would be a better user experience than just an attribute.

Or
```py
    BMWBinarySensorEntityDescription(
        key="lids",
        translation_key="lids",
        device_class=BinarySensorDeviceClass.OPENING,
        # device class opening: On means open, Off means closed
        value_fn=lambda v: not v.doors_and_windows.all_lids_closed,
        attr_fn=lambda v, u: {
            lid.name: lid.state.value for lid in v.doors_and_windows.lids
        },
```
Why don't we create an entity per lid?

Reply:
Thinking about it even more, probably we can do it for most I would disable them all by default though (but maybe @gerard33 has some other ideas here).

In order of appearance in `SENSOR_TYPES`:
- `lids`: as a user, I'd rather now IF a door is open, not which. Could add them as hidden binary_sensors per lid
- `windows`: same reasoning as above
- `door_lock_state`: Agree with you, we do something similar for `charging_state`
- `condition_based_services`: we could create two sensors per available condition based service (so 10*2 = 20) to track 
  1. remaining km
  2. timestamp when the car needs service.
- `check_control_message`: no idea how a sensor would make sense here, I would leave it (except for a enum sensor that is 99% of time empty/none and displays the first, but not all, actual messages)

Reply:
> as a user, I'd rather now IF a door is open, not which. Could add them as hidden binary_sensors per lid

But with that reasoning we could also remove the attributes. The main reason I am also asking this is that changes like this clean up the state machine a bit

Oh so `condition_based_services` is about when the car needs service for that specific field? I think that could work.

`check_control_message`, the API is polling right? Otherwise it would've been a nice event entity. But its fine leaving this then.

But again, these are not blockers, just general remarks. If a new integration would try to add this I would've asked the same questions :)

Reply:
I think I'll work on these in a separate PR. Should these still be added to the quality scale somewhere (like the other non-blocking suggestions)?

We could use the 5 minute push for`check_control_message`. 
However, I don't think event is the right platform, as it doesn't clear its state:
> Home Assistant will keep track of the last event that was emitted and will show that as the current state of the entity.

Reply:
I think it makes sense to migrate the attributes to sensors as much as possible as it gives easier access to the actual state. And most of them can indeed be disabled by default.

Reply:
Event entities work well for events as it shows "event x was 7 days ago". But they don't work well in polling integrations as they aren't accurate